docs/blog/20/3-mii-signal-bug-in-xilinx-fpga.md

```diff
@@ -0,0 +1,237 @@
+    to do with `ENET0_GMII_TX_*`. They do not affect TX side in any ways.
+
+As can be seen from the figure, there are just four signals that should be taken
+into consideration. These signals are `ENET0_GMII_TX_EN`, `ENET0_GMII_TX_ER`,
+`ENET0_GMII_TXD`, `ENET0_GMII_TX_CLK`. Because the speed is chosen as 100
+Mbits/s, 25 MHz clock is generated via `ZYNQ Block` and connected to
+`ENET0_GMII_TX_CLK`. Other **TX signals**** are connected to `ILA Block` to make
+some chipscope tests.
+
+Moreover, I have utilized **Linux Driver** for setupping transmission with
+**ZYNQ GEM**. In a more understandable language, I have used Linux Driver so
+that **ARM side** can generate Ethernet packets and send them via MII interface.
+
+As a result, after building this, generating bitstream and programming device, I
+have started to observe MII signals and comprehend how they are sent via ILA.
+However, in some packets, I have found a **bug** that is explained in the next
+part.
+
+## **THE BUG**
+
+According to MII Protocol Specification, during `TX_EN` signal assertion,
+receiver side takes the nibbles (for standard MII). When `TX_EN` signal goes
+low, receiver side takes the entire packet until that time and makes some
+process (such as calculating **CRC** and comparing with the received one) on the
+packet during **interpacket gap time**. Therefore, when `TX_EN` signal is not
+high after some transmission, receiver side understands that the packet is
+finished and it is time to make some processes.
+
+However, in my architecture, sometimes, in the middle of packet, `TX_EN` signal
+goes low for a single clock cycle and then again goes high. So, it does not stay
+low for **24 clock cycles (12 octets)** after it goes low. Also, so that I am
+sure that it is a bug, I have written a module that writes MII data to **BRAM**.
+For bug case, I have calculated CRC without including the case that `TX_EN` goes
+low and I have compared the calculated CRC with the received CRC. They were
+exactly the same.
+
+This kind of bug can result in some errors in receiver side because people write
+codes according to the specifications and if you do not satisfy them, there
+always occur some errors and bugs. I have put some screenshoots from **ILA** to
+comprehend the problem more easily:
+
+![ILA Problem-First Image](img/3-ila_problem_v1.png)
+
+![ILA Problem-First Image](img/3-ila_problem_v2.png)
+
+As can be inferred from the pictures, somewhere in the middle of the entire MII
+packet, `TX_EN` goes low. We can say that a packet transmission has finished and
+a new one has started but they are not two seperate packets according to my CRC
+calculations. Because they unite a packet, `TX_EN` should not be disasserted
+even for a single clock cycle.
+
+<!--markdownlint-disable MD034-->
+Finally, for this bug, I have posted a detailed question on [Xilinx Forum]
+(https://forums.xilinx.com/t5/Ethernet/MII-Signals-Problem-on-ILA/m-p/1126290#M28573)
```

Comment:
Bu link iki satıra bölündüğü için çalışmıyor, tek satır olmalı. Lokalde denersen görebilirsin. MD013 açısından linkin uzun olması tek satırda sorun değil.
```suggestion
[Xilinx Forum](https://forums.xilinx.com/t5/Ethernet/MII-Signals-Problem-on-ILA/m-p/1126290#M28573)
```